File: src/main/java/no/fint/provider/springer/model/TimeplanRepository.java
package no.fint.provider.springer.model;

import lombok.extern.slf4j.Slf4j;
import no.fint.event.model.Event;
import no.fint.model.resource.FintLinks;
import no.fint.model.resource.utdanning.timeplan.FagResource;
import no.fint.model.resource.utdanning.timeplan.RomResource;
import no.fint.model.resource.utdanning.timeplan.TimeResource;
import no.fint.model.resource.utdanning.timeplan.UndervisningsgruppeResource;
import no.fint.model.utdanning.timeplan.TimeplanActions;
import no.fint.provider.springer.storage.SpringerRepository;
import org.springframework.stereotype.Repository;

import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static no.fint.model.utdanning.timeplan.TimeplanActions.*;

@Slf4j
@Repository
public class TimeplanRepository extends SpringerRepository {

    @Override
    public void accept(Event<FintLinks> response) {
        switch (TimeplanActions.valueOf(response.getAction())) {
            case GET_ALL_FAG:
                query(FagResource.class, response);
                break;
            case GET_ALL_ROM:
                query(RomResource.class, response);
                break;
            case GET_ALL_TIME:
                query(TimeResource.class, response);
                break;
            case GET_ALL_UNDERVISNINGSGRUPPE:
                query(UndervisningsgruppeResource.class, response);
                break;
        }
    }

    @Override
    public Set<String> actions() {
        return Stream
            .of(GET_ALL_FAG, GET_ALL_ROM, GET_ALL_TIME, GET_ALL_UNDERVISNINGSGRUPPE)
            .map(Enum::name)
            .collect(Collectors.toSet());
    }
}

File: src/main/kotlin/no/fint/provider/springer/model/TimeplanRepository.kt
package no.fint.provider.springer.model

import no.fint.event.model.Event
import no.fint.event.model.ResponseStatus
import no.fint.event.model.Status
import no.fint.provider.springer.storage.SpringerRepository
import no.fint.provider.springer.storage.Wrapper
import no.novari.fint.model.resource.FintLinks
import no.novari.fint.model.resource.utdanning.timeplan.FagResource
import no.novari.fint.model.resource.utdanning.timeplan.RomResource
import no.novari.fint.model.resource.utdanning.timeplan.TimeResource
import no.novari.fint.model.resource.utdanning.timeplan.UndervisningsgruppeResource
import no.novari.fint.model.resource.utdanning.timeplan.UndervisningsgruppemedlemskapResource
import no.novari.fint.model.utdanning.timeplan.TimeplanActions
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.stereotype.Repository

@Repository
class TimeplanRepository(
    mongoTemplate: MongoTemplate,
    wrapper: Wrapper
) : SpringerRepository(mongoTemplate, wrapper) {

    override fun accept(response: Event<FintLinks>) {
        when (TimeplanActions.valueOf(response.action)) {
            TimeplanActions.GET_ALL_FAG -> query(FagResource::class.java, response)
            TimeplanActions.GET_ALL_ROM -> query(RomResource::class.java, response)
            TimeplanActions.GET_ALL_TIME -> query(TimeResource::class.java, response)
            TimeplanActions.GET_ALL_UNDERVISNINGSGRUPPE -> query(UndervisningsgruppeResource::class.java, response)
            TimeplanActions.GET_ALL_UNDERVISNINGSGRUPPEMEDLEMSKAP -> query(UndervisningsgruppemedlemskapResource::class.java, response)
            else -> invalid(response)
        }
    }

    private fun invalid(response: Event<FintLinks>) {
        response.status = Status.ADAPTER_REJECTED
        response.responseStatus = ResponseStatus.REJECTED
        response.statusCode = "INVALID_ACTION"
        response.message = "Invalid action"
    }

    override fun actions(): Set<String> =
        TimeplanActions.entries
            .map { it.name }
            .filter { it.startsWith("GET_ALL_") }
            .toSet()
}
